File: C++/sorting/quicksort.cpp
#include "quicksort.h"

void Quicksort::Sort(std::vector<int>& arr) {
  Recursive(arr, 0, arr.size() - 1);
}

void Quicksort::Recursive(std::vector<int>& arr, int low, int high) {
  if (low < high) {
    int pi = Partition(arr, low, high);

    Recursive(arr, low, pi - 1);
    Recursive(arr, pi + 1, high);
  }
}

int Quicksort::Partition(std::vector<int>& arr, int low, int high) {
  int pivot = arr[high];
  int i = low - 1;

  for (int j = low; j < high; ++j) {
    if (arr[j] < pivot) {
      std::swap(arr[++i], arr[j]);
    }
  }
  std::swap(arr[++i], arr[high]);
  return i;
}

File: Java/sorting/Quicksort.java
package sorting;

import java.util.List;

public class Quicksort {
  public static void sort(List<Integer> arr) {
    quicksort(arr, 0, arr.size() - 1);
  }

  private static void quicksort(List<Integer> arr, int left, int right) {
    if (right > left) {
      int pi = partition(arr, left, right);

      quicksort(arr, left, pi - 1);
      quicksort(arr, pi + 1, right);
    }    
  }

  private static int partition(List<Integer> arr, int left, int right) {
    int pivot = arr.get(right);
    int i = left - 1;

    for (int j = left; j < right; j++) {
      if (arr.get(j) < pivot) {
        swap(arr, j, ++i);
      }
    }
    swap(arr, ++i, right);
    return i;
  }

  private static void swap(List<Integer> arr, int left, int right) {
    int temp = arr.get(left);
    arr.set(left, arr.get(right));
    arr.set(right, temp);
  }
}
